clientv3/balancer.go

```diff
@@ -15,39 +15,108 @@
 package clientv3
 
 import (
+	"math/rand"
 	"net/url"
 	"strings"
-	"sync/atomic"
+	"sync"
+	"time"
 
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
 )
 
+// pinWait is the amount of time to wait before pinning a new address
+var defaultPinWait = 50 * time.Millisecond
+
 // simpleBalancer does the bare minimum to expose multiple eps
 // to the grpc reconnection code path
 type simpleBalancer struct {
-	// eps are the client's endpoints stripped of any URL scheme
-	eps     []string
-	ch      chan []grpc.Address
-	numGets uint32
+	// addrs are the client's endpoints for grpc
+	addrs []grpc.Address
+
+	ch chan []grpc.Address
+
+	// upc receives a message whenever up is posted
+	upc chan struct{}
+
+	// pinWait is the amount of time to wait before trying a new pin
+	pinWait time.Duration
+
+	// mu protects upEps, getIdx, pinAddr, and pinTime
+	mu sync.Mutex
+	// upEps contains the endpoints grpc says are up
+	upEps map[string]struct{}
+	// getIdx chooses the next endpoint
+	getIdx uint32
+	// pinAddr is address of the pinned connection
+	pinAddr *grpc.Address
+	// pinTime is the time when pinAddr was set
+	pinTime time.Time
 }
 
-func newSimpleBalancer(eps []string) grpc.Balancer {
+func newSimpleBalancer(eps []string, timeout time.Duration) *simpleBalancer {
 	ch := make(chan []grpc.Address, 1)
 	addrs := make([]grpc.Address, len(eps))
 	for i := range eps {
 		addrs[i].Addr = getHost(eps[i])
 	}
 	ch <- addrs
-	return &simpleBalancer{eps: eps, ch: ch}
+	pinWait := timeout
+	if pinWait == 0 {
+		pinWait = defaultPinWait
+	}
+	return &simpleBalancer{
+		addrs:   addrs,
+		ch:      ch,
+		upc:     make(chan struct{}, 1),
+		pinWait: pinWait,
+		upEps:   make(map[string]struct{}),
+		// randomize initial getIdx to distribute connections
+		getIdx: uint32(rand.Intn(len(eps))),
+	}
+}
+
+func (b *simpleBalancer) Start(target string) error { return nil }
+
+func (b *simpleBalancer) Up(addr grpc.Address) func(error) {
+	b.mu.Lock()
+	b.upEps[addr.Addr] = struct{}{}
+	b.mu.Unlock()
+	// notify client that a connection is up
+	select {
+	case b.upc <- struct{}{}:
+	default:
+	}
+	return func(error) {
+		b.mu.Lock()
+		delete(b.upEps, addr.Addr)
+		if b.pinAddr != nil && b.pinAddr.Addr == addr.Addr {
+			b.pinAddr = nil
+		}
+		b.mu.Unlock()
+	}
 }
 
-func (b *simpleBalancer) Start(target string) error        { return nil }
-func (b *simpleBalancer) Up(addr grpc.Address) func(error) { return func(error) {} }
 func (b *simpleBalancer) Get(ctx context.Context, opts grpc.BalancerGetOptions) (grpc.Address, func(), error) {
-	v := atomic.AddUint32(&b.numGets, 1)
-	ep := b.eps[v%uint32(len(b.eps))]
-	return grpc.Address{Addr: getHost(ep)}, func() {}, nil
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.pinAddr != nil {
+		if _, ok := b.upEps[b.pinAddr.Addr]; ok || time.Since(b.pinTime) < b.pinWait {
```

Comment:
OK, this code is bad; the dial timeout is already in the grpc dialer and there should only be one connecting / up endpoint at a time. I can simplify it.